File: lib/models/supabase/journey_progress.py
import json
from datetime import datetime
from typing import Optional
from lib.models.supabase.supabase_model import SupabaseModel
from uuid import UUID
from pydantic import Field, Json, field_validator
from supabase.client import AsyncClient


class JourneyProgressModel(SupabaseModel):
    TABLE_NAME: str = "journey_progress"
    id: Optional[UUID] = Field(default=None)
    journey_id: UUID
    journey_version_id: UUID
    assigned_at: Optional[datetime] = Field(default=None)
    metadata: Optional[Json] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: UUID
    started_at: Optional[datetime] = Field(default=None)
    completed_at: Optional[datetime] = Field(default=None)

    @field_validator("metadata", mode="before")
    def validate_metadata(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        return v


class JourneyProgressLLMConversationMessagesModel(SupabaseModel):
    TABLE_NAME: str = "journey_progress_llm_conversation_messages"
    journey_item_progress_id: UUID
    message_id: UUID
    conversation_id: UUID
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: UUID

    async def save_to_supabase(
        self,
        supabase: AsyncClient,
        on_conflict=["journey_item_progress_id", "message_id"],
    ):
        await super().save_to_supabase(supabase, on_conflict)

    async def fetch_from_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "journey_item_progress_id": self.journey_item_progress_id,
                "message_id": self.message_id,
            }
        return await super().fetch_from_supabase(
            supabase, value=value, id_field_name=id_field_name
        )

    async def exists_in_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "journey_item_progress_id": self.journey_item_progress_id,
                "message_id": self.message_id,
            }
        return await super().exists_in_supabase(
            supabase, value=value, id_field_name=id_field_name
        )

    async def delete_from_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "journey_item_progress_id": self.journey_item_progress_id,
                "message_id": self.message_id,
            }
        return await super().delete_from_supabase(
            supabase, value=value, id_field_name=id_field_name
        )


class JourneyProgressLLMConversationsModel(SupabaseModel):
    TABLE_NAME: str = "journey_progress_llm_conversations"
    progress_id: UUID
    conversation_id: UUID
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: UUID

    async def save_to_supabase(
        self, supabase: AsyncClient, on_conflict=["progress_id", "conversation_id"]
    ):
        await super().save_to_supabase(supabase, on_conflict)

    async def fetch_from_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "progress_id": self.progress_id,
                "conversation_id": self.conversation_id,
            }
        return await super().fetch_from_supabase(
            supabase, value=value, id_field_name=id_field_name
        )

    async def exists_in_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "progress_id": self.progress_id,
                "conversation_id": self.conversation_id,
            }
        return await super().exists_in_supabase(
            supabase, value=value, id_field_name=id_field_name
        )

    async def delete_from_supabase(
        self, supabase: AsyncClient, value=None, id_field_name=None
    ):
        if value is None:
            value = {
                "progress_id": self.progress_id,
                "conversation_id": self.conversation_id,
            }
        return await super().delete_from_supabase(
            supabase, value=value, id_field_name=id_field_name
        )


class JourneyItemProgressModel(SupabaseModel):
    TABLE_NAME: str = "journey_item_progress"
    id: Optional[UUID] = Field(default=None)
    journey_progress_id: UUID
    journey_item_id: UUID
    journey_item_version_id: UUID
    data: Optional[Json] = Field(default=None)
    created_at: Optional[datetime] = Field(default=None)
    updated_at: Optional[datetime] = Field(default=None)
    owner_id: Optional[UUID] = Field(default=None)
    organization_id: UUID
    started_at: Optional[datetime] = Field(default=None)
    completed_at: Optional[datetime] = Field(default=None)

    @field_validator("data", mode="before")
    def validate_data(cls, v):
        if isinstance(v, str):
            return json.loads(v)
        return v

File: lib_js/models/supabase/journeyProgress.js
import { SupabaseModel } from "./supabaseModel";
import { v4 as uuidv4 } from "uuid";

// Define JourneyProgressModel
export class JourneyProgressModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress";

    constructor(args) {
        super();
        const {
            id = null,
            journeyId,
            journeyVersionId,
            assignedAt = null,
            metadata = null,
            createdAt = null,
            updatedAt = null,
            ownerId = null,
            organizationId = null,
            startedAt = null,
            completedAt = null
        } = args;
        this.attributes = {
            id: {
                value: id || uuidv4(),
                type: "uuid",
                required: false,
                dbColumn: "id"
            },
            journeyId: {
                value: journeyId,
                type: "uuid",
                required: true,
                dbColumn: "journey_id"
            },
            journeyVersionId: {
                value: journeyVersionId,
                type: "uuid",
                required: true,
                dbColumn: "journey_version_id"
            },
            assignedAt: {
                value: assignedAt,
                type: "date",
                required: false,
                dbColumn: "assigned_at"
            },
            metadata: {
                value: metadata,
                type: "json",
                required: false,
                dbColumn: "metadata"
            },
            createdAt: {
                value: createdAt,
                type: "date",
                required: false,
                dbColumn: "created_at"
            },
            updatedAt: {
                value: updatedAt,
                type: "date",
                required: false,
                dbColumn: "updated_at"
            },
            ownerId: {
                value: ownerId,
                type: "uuid",
                required: false,
                dbColumn: "owner_id"
            },
            organizationId: {
                value: organizationId,
                type: "uuid",
                required: false,
                dbColumn: "organization_id"
            },
            startedAt: {
                value: startedAt,
                type: "date",
                required: false,
                dbColumn: "started_at"
            },
            completedAt: {
                value: completedAt,
                type: "date",
                required: false,
                dbColumn: "completed_at"
            }
        };
    }
}

// Define JourneyProgressLLMConversationMessagesModel
export class JourneyProgressLLMConversationMessagesModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress_llm_conversation_messages";

    constructor(args) {
        super();
        const {
            journeyItemProgressId,
            messageId,
            conversationId,
            createdAt = null,
            updatedAt = null,
            ownerId = null,
            organizationId = null
        } = args;
        this.attributes = {
            journeyItemProgressId: {
                value: journeyItemProgressId,
                type: "uuid",
                required: true,
                dbColumn: "journey_item_progress_id"
            },
            messageId: {
                value: messageId,
                type: "uuid",
                required: true,
                dbColumn: "message_id"
            },
            conversationId: {
                value: conversationId,
                type: "uuid",
                required: true,
                dbColumn: "conversation_id"
            },
            createdAt: {
                value: createdAt,
                type: "date",
                required: false,
                dbColumn: "created_at"
            },
            updatedAt: {
                value: updatedAt,
                type: "date",
                required: false,
                dbColumn: "updated_at"
            },
            ownerId: {
                value: ownerId,
                type: "uuid",
                required: false,
                dbColumn: "owner_id"
            },
            organizationId: {
                value: organizationId,
                type: "uuid",
                required: false,
                dbColumn: "organization_id"
            }
        };
    }

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["journeyItemProgressId", "messageId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instances,
        onConflict = ["journeyItemProgressId", "messageId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(
            supabase,
            instances,
            onConflict,
            idColumn
        );
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId:
                    value.attributes.journeyItemProgressId.value,
                messageId: value.attributes.messageId.value
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId:
                    value.attributes.journeyItemProgressId.value,
                messageId: value.attributes.messageId.value
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                journeyItemProgressId:
                    value.attributes.journeyItemProgressId.value,
                messageId: value.attributes.messageId.value
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

// Define JourneyProgressLLMConversationsModel
export class JourneyProgressLLMConversationsModel extends SupabaseModel {
    static TABLE_NAME = "journey_progress_llm_conversations";

    constructor(args) {
        super();
        const {
            progressId,
            conversationId,
            createdAt = null,
            updatedAt = null,
            ownerId = null,
            organizationId = null
        } = args;
        this.attributes = {
            progressId: {
                value: progressId,
                type: "uuid",
                required: true,
                dbColumn: "progress_id"
            },
            conversationId: {
                value: conversationId,
                type: "uuid",
                required: true,
                dbColumn: "conversation_id"
            },
            createdAt: {
                value: createdAt,
                type: "date",
                required: false,
                dbColumn: "created_at"
            },
            updatedAt: {
                value: updatedAt,
                type: "date",
                required: false,
                dbColumn: "updated_at"
            },
            ownerId: {
                value: ownerId,
                type: "uuid",
                required: false,
                dbColumn: "owner_id"
            },
            organizationId: {
                value: organizationId,
                type: "uuid",
                required: false,
                dbColumn: "organization_id"
            }
        };
    }

    static async saveToSupabase(
        supabase,
        instance,
        onConflict = ["progressId", "conversationId"]
    ) {
        return super.saveToSupabase(supabase, instance, onConflict);
    }

    static async upsertToSupabase(
        supabase,
        instance,
        onConflict = ["progressId", "conversationId"],
        idColumn = null
    ) {
        return super.upsertToSupabase(supabase, instance, onConflict, idColumn);
    }

    static async fetchFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId.value,
                conversationId: value.attributes.conversationId.value
            };
        }
        return super.fetchFromSupabase(supabase, value, idColumn);
    }

    static async existsInSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId.value,
                conversationId: value.attributes.conversationId.value
            };
        }
        return super.existsInSupabase(supabase, value, idColumn);
    }

    static async deleteFromSupabase(supabase, value = null, idColumn = null) {
        if (value instanceof this) {
            value = {
                progressId: value.attributes.progressId.value,
                conversationId: value.attributes.conversationId.value
            };
        }
        return super.deleteFromSupabase(supabase, value, idColumn);
    }
}

// Define JourneyItemProgressModel
export class JourneyItemProgressModel extends SupabaseModel {
    static TABLE_NAME = "journey_item_progress";

    constructor(args) {
        super();
        const {
            id = null,
            journeyProgressId,
            journeyItemId,
            journeyItemVersionId,
            data = null,
            createdAt = null,
            updatedAt = null,
            ownerId = null,
            organizationId = null,
            startedAt = null,
            completedAt = null
        } = args;
        this.attributes = {
            id: {
                value: id || uuidv4(),
                type: "uuid",
                required: false,
                dbColumn: "id"
            },
            journeyProgressId: {
                value: journeyProgressId,
                type: "uuid",
                required: true,
                dbColumn: "journey_progress_id"
            },
            journeyItemId: {
                value: journeyItemId,
                type: "uuid",
                required: true,
                dbColumn: "journey_item_id"
            },
            journeyItemVersionId: {
                value: journeyItemVersionId,
                type: "uuid",
                required: true,
                dbColumn: "journey_item_version_id"
            },
            data: {
                value: data,
                type: "json",
                required: false,
                dbColumn: "data"
            },
            createdAt: {
                value: createdAt,
                type: "date",
                required: false,
                dbColumn: "created_at"
            },
            updatedAt: {
                value: updatedAt,
                type: "date",
                required: false,
                dbColumn: "updated_at"
            },
            ownerId: {
                value: ownerId,
                type: "uuid",
                required: false,
                dbColumn: "owner_id"
            },
            organizationId: {
                value: organizationId,
                type: "uuid",
                required: false,
                dbColumn: "organization_id"
            },
            startedAt: {
                value: startedAt,
                type: "date",
                required: false,
                dbColumn: "started_at"
            },
            completedAt: {
                value: completedAt,
                type: "date",
                required: false,
                dbColumn: "completed_at"
            }
        };
    }
}
